docs/book/v2/routing-middleware.md

```diff
@@ -0,0 +1,141 @@
+            ],
+            'instance-handler' => [
+                'type' => Literal::class,
+                'options' => [
+                    'route' => '/instance-handler',
+                    'defaults' => [
+                        'controller' => PipeSpec::class,
+                        'middleware' => new class implements RequestHandlerInterface {
+                            public function handle(
+                                ServerRequestInterface $request
+                            ): ResponseInterface {
+                                //  
+                            }
+                        },
+                    ],
+                ],
+            ],
+            'instance-pipespec-preferred' => [
+                'type' => Literal::class,
+                'options' => [
+                    'route' => '/instance-pipespec-preferred',
+                    'defaults' => [
+                        'controller' => PipeSpec::class,
+                        'middleware' => new PipeSpec(
+                            'container_id_for_middleware',
+                            'container_id_for_another_middleware',
+                            'container_id_for_handler',
+                        ),
+                    ],
+                ],
+            ],
+            'instance-pipespec-dev-env' => [
+                'type' => Literal::class,
+                'options' => [
+                    'route' => '/instance-pipespec-dev-env',
+                    'defaults' => [
+                        'controller' => PipeSpec::class,
+                        'middleware' => new PipeSpec(
+                            'container_id_for_middleware',
+                            new class implements MiddlewareInterface {
+                                public function process(
+                                    ServerRequestInterface $request,
+                                    RequestHandlerInterface $handler
+                                ): ResponseInterface {
+                                    return $handler->handle($request);
+                                }
+                            },
+                            static function (
+                                ServerRequestInterface $request,
+                                RequestHandlerInterface $handler
+                            ): ResponseInterface {
+                                return $handler->handle($request);
+                            },
+                            'container_id_for_handler',
+                        ),
+                    ],
+                ],
+            ],
+        ],
+    ],
+];
+```
+
+Please note that if you pass instances of middleware or handlers directly, it is not compatible with config caching.
+You should sue that for development only.
+The recommended variant ist passing container managed middleware, i.e. container service strings
+like `my-awesome-middleware` or `SomeRequestHandler::class`.
```

Comment:
I think the class name should be recommended because a simple string leads only to spelling errors and can not be checked via IDE.